client/rpcproxy/rpcproxy.go

```diff
@@ -0,0 +1,765 @@
+
+// saveServerList is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
+func (p *RpcProxy) saveServerList(l serverList) {
+	p.activatedList.Store(l)
+}
+
+func (p *RpcProxy) LeaderAddr() string {
+	p.listLock.Lock()
+	defer p.listLock.Unlock()
+	return p.leaderAddr
+}
+
+// New is the only way to safely create a new RpcProxy.
+func New(logger *log.Logger, shutdownCh chan struct{}, configInfo NomadConfigInfo, connPoolPinger Pinger) (p *RpcProxy) {
+	p = new(RpcProxy)
+	p.logger = logger
+	p.configInfo = configInfo         // can't pass *nomad.Client: import cycle
+	p.connPoolPinger = connPoolPinger // can't pass *nomad.ConnPool: import cycle
+	p.rebalanceTimer = time.NewTimer(clientRPCMinReuseDuration)
+	p.shutdownCh = shutdownCh
+
+	l := serverList{}
+	l.L = make([]*ServerEndpoint, 0)
+	p.saveServerList(l)
+	return p
+}
+
+// NotifyFailedServer marks the passed in server as "failed" by rotating it
+// to the end of the server list.
+func (p *RpcProxy) NotifyFailedServer(s *ServerEndpoint) {
+	l := p.getServerList()
+
+	// If the server being failed is not the first server on the list,
+	// this is a noop.  If, however, the server is failed and first on
+	// the list, acquire the lock, retest, and take the penalty of moving
+	// the server to the end of the list.
+
+	// Only rotate the server list when there is more than one server
+	if len(l.L) > 1 && l.L[0] == s &&
+		// Use atomic.CAS to emulate a TryLock().
+		atomic.CompareAndSwapInt32(&p.notifyFailedBarrier, 0, 1) {
+		defer atomic.StoreInt32(&p.notifyFailedBarrier, 0)
+
+		// Grab a lock, retest, and take the hit of cycling the first
+		// server to the end.
+		p.listLock.Lock()
+		defer p.listLock.Unlock()
+		l = p.getServerList()
+
+		if len(l.L) > 1 && l.L[0] == s {
+			l.L = l.cycleServer()
+			p.saveServerList(l)
+		}
+	}
+}
+
+func (p *RpcProxy) NumNodes() int {
+	return p.numNodes
+}
+
+// NumServers takes out an internal "read lock" and returns the number of
+// servers.  numServers includes both healthy and unhealthy servers.
+func (p *RpcProxy) NumServers() int {
+	l := p.getServerList()
+	return len(l.L)
+}
+
+// RebalanceServers shuffles the list of servers on this agent.  The server
+// at the front of the list is selected for the next RPC.  RPC calls that
+// fail for a particular server are rotated to the end of the list.  This
+// method reshuffles the list periodically in order to redistribute work
+// across all known Nomad servers (i.e. guarantee that the order of servers
+// in the server list is not positively correlated with the age of a server
+// in the Nomad cluster).  Periodically shuffling the server list prevents
+// long-lived clients from fixating on long-lived servers.
+//
+// Unhealthy servers are removed from the server list during the next client
+// heartbeat.  Before the newly shuffled server list is saved, the new remote
+// endpoint is tested to ensure its responsive.
+func (p *RpcProxy) RebalanceServers() {
```

Comment:
Here and the next function change the variable names to be more verbose in their meaning